plugin/hover.py

```diff
@@ -120,8 +121,16 @@ def run_async() -> None:
 
     def request_symbol_hover_async(self, listener: AbstractViewListener, point: int) -> None:
         hover_promises = []  # type: List[Promise[ResolvedHover]]
-        document_position = text_document_position_params(self.view, point)
         for session in listener.sessions_async('hoverProvider'):
+            document_position = text_document_position_params(self.view, point)
+            range_hover_provider = session.get_capability('experimental.rangeHoverProvider')
+            if range_hover_provider:
+                region = first_selection_region(self.view)
+                if region is not None:
+                    if region.contains(point):
+                        # hovering selection or only text was selected
+                        document_position.pop('position', None)
+                        document_position['range'] = region_to_range(self.view, region).to_lsp()
```

Comment:
This looks very messy. You could make it simpler and fix types with something like:

```suggestion
            params = self._create_hover_request(session, point)
            hover_promises.append(session.send_request_task(Request("textDocument/hover", params, self.view)))
```

and define

```py
    def _create_hover_request(
        self, session: Session, point: int
    ) -> Union[TextDocumentPositionParams, ExperimentalTextDocumentRangeParams]:
        if session.get_capability('experimental.rangeHoverProvider'):
            region = first_selection_region(self.view)
            if region is not None and region.contains(point):
                return text_document_range_params(self.view, region)
        return text_document_position_params(self.view, point)
```

I would also create proper type for the `text_document_range_params`:

```py
ExperimentalTextDocumentRangeParams = TypedDict('ExperimentalTextDocumentRangeParams', {
    'textDocument': TextDocumentIdentifier,
    'range': RangeLsp,
}, total=True)
```